File: common/src/main/java/dev/mayaqq/estrogen/client/cosmetics/ui/CosmeticsLoginScreen.java
package dev.mayaqq.estrogen.client.cosmetics.ui;

import com.simibubi.create.foundation.gui.ScreenOpener;
import com.simibubi.create.foundation.gui.Theme;
import dev.mayaqq.estrogen.client.cosmetics.service.CosmeticsApi;
import dev.mayaqq.estrogen.client.cosmetics.ui.widgets.UIComponents;
import net.minecraft.client.Minecraft;
import net.minecraft.client.gui.GuiGraphics;
import net.minecraft.client.gui.screens.Screen;
import net.minecraft.network.chat.Component;
import net.minecraft.util.FormattedCharSequence;

public class CosmeticsLoginScreen extends BaseCosmeticsScreen {

    private Component info;

    public CosmeticsLoginScreen(Screen parent) {
        super(parent);
    }

    @Override
    protected void contentInit(int x, int y, int width, int height) {
        this.claimButton.asDisabled();

        int loginButtonX = x + width / 2 - 50;
        int loginButtonY = (int) (y + height * 0.8f - 10);

        addRenderableWidget(UIComponents.textButton(CosmeticUI.LOGIN_BUTTON)
                .withPosition(loginButtonX, loginButtonY)
                .withBounds(100, 20)
                .withCallback(this::login)
        );
    }

    @Override
    protected void renderWindow(GuiGraphics graphics, int mouseX, int mouseY, float partialTicks) {
        super.renderWindow(graphics, mouseX, mouseY, partialTicks);

        int x = (int) (this.width * 0.25f) + PADDING * 2;
        int y = PADDING * 3;
        int width = this.width - x - PADDING;

        int infoX = x + width / 2;
        int infoY = y;

        for (FormattedCharSequence text : font.split(CosmeticUI.LOGIN_DESCRIPTION, width)) {
            graphics.drawCenteredString(font, text, infoX, infoY, Theme.i(Theme.Key.BUTTON_IDLE));
            infoY += 10;
        }

        if (this.info != null) {
            infoY += 10;
            graphics.drawCenteredString(font, this.info, infoX, infoY, Theme.i(Theme.Key.BUTTON_FAIL));
        }
    }

    public void login() {
        this.info = CosmeticUI.getLoginMessage(null);

        CosmeticsApi.login().thenAcceptAsync(status -> {
            if (status == CosmeticsApi.StatusCode.OK) {
                this.info = CosmeticUI.getCosmeticsMessage(null);
                CosmeticsApi.getCosmetics().thenAcceptAsync(cosmeticsStatus -> {
                    if (cosmeticsStatus == CosmeticsApi.StatusCode.OK) {
                        Minecraft.getInstance().tell(() -> {
                            Screen screen = Minecraft.getInstance().screen;
                            if (screen != null) {
                                screen.onClose();
                                ScreenOpener.open(new CosmeticsScreen(null));
                            }
                        });
                    } else {
                        this.info = CosmeticUI.getCosmeticsMessage(cosmeticsStatus);
                    }
                });
            } else {
                this.info = CosmeticUI.getLoginMessage(status);
            }
        });
    }
}

File: src/main/kotlin/dev/mayaqq/estrogen/client/content/screen/cosmetics/CosmeticsLoginScreen.kt
package dev.mayaqq.estrogen.client.content.screen.cosmetics

import dev.mayaqq.cynosure.helpers.McClient
import dev.mayaqq.cynosure.helpers.McFont
import dev.mayaqq.cynosure.helpers.setScreenAsync
import dev.mayaqq.cynosure.text.TextUtils.splitLines
import dev.mayaqq.cynosure.text.unaryMinus
import dev.mayaqq.estrogen.client.content.screen.EstrogenButton
import dev.mayaqq.estrogen.client.content.screen.EstrogenMenuScreen
import dev.mayaqq.estrogen.client.cosmetics.CosmeticAPI
import dev.mayaqq.estrogen.client.cosmetics.StatusCode
import net.minecraft.client.gui.GuiGraphics
import net.minecraft.client.gui.screens.Screen
import net.minecraft.network.chat.Component


class CosmeticsLoginScreen(previous: Screen?) : CosmeticsBaseScreen(previous, -"gui.estrogen.cosmetics.title", false) {

    var info: Component? = null

    val bLogin = EstrogenButton.Builder(EstrogenButton.TextRenderer(-"gui.estrogen.cosmetics.login")) {
        this.info = getLoginMessage(null)
        CosmeticAPI.login().thenAcceptAsync { status ->
            if (status === StatusCode.OK) {
                this.info = getCosmeticsMessage(null)
                CosmeticAPI.getCosmetics().thenAcceptAsync { cosmeticsStatus ->
                    if (cosmeticsStatus === StatusCode.OK) {
                        McClient.setScreenAsync{ CosmeticsSelectionScreen(previous) }
                    } else {
                        this.info = getCosmeticsMessage(cosmeticsStatus)
                    }
                }
            } else {
                this.info = getLoginMessage(status)
            }
        }
    }.color(EstrogenMenuScreen.transWhite)

    override fun baseInit() {
        super.baseInit()
        val fakeThirdWidth = (width - 30) / 3
        45 + ((height - 90) / 2)
        bLogin.bounds(20 + fakeThirdWidth + 80, height - 35, fakeThirdWidth * 2 - 160, 25).buildAndAdd()
    }

    override fun baseRender(graphics: GuiGraphics, mouseX: Int, mouseY: Int, partialTick: Float) {
        val fakeThirdWidth = (width - 30) / 3
        val x = (fakeThirdWidth * 2) + 20
        var y = ((height - 90) / 2) - (McFont.lineHeight * 3)
        (-"gui.estrogen.cosmetics.login.description").splitLines().forEach { line ->
            graphics.drawCenteredString(McFont, line, x, y, 0xFFFFFF)
            y += McFont.lineHeight
        }
        info?.let {
            y += McFont.lineHeight
            graphics.drawCenteredString(font, it, x, y, 0xFFFFFF)
        }
    }
}